### PR TITLE
Remove state vector support from `math/quantum.py`

### PR DESCRIPTION
**Pull request by eddddddy**
**Context:**
Part of the deprecations/removals for 0.32. This PR is a followup to the deprecation of state-vector arguments to most functions in `pennylane/math/quantum.py`.

**Description of the Change:**
Remove state-vector support from `purity`, `vn_entropy`, `mutual_info`, `fidelity`, `relative_entropy`, and `max_entropy`. 

Also add another function `fidelity_statevector` which computes the fidelity given two state-vectors, since this is a widely-used case that could benefit from a separate function.

### CODE REVIEW

File: pennylane/math/quantum.py [hunk 1]
```diff
@@ -859,74 +772,91 @@ def fidelity(state0, state1, check_state=False, c_dtype="complex128"):
...
+    if not is_abstract(state0):
+        state1 = cast_like(state1, state0)
+
+    if check_state:
+        _check_state_vector(state0)
+        _check_state_vector(state1)
+
+    # Get dimension of the quantum system and reshape
+    num_indices0 = int(np.log2(qml.math.shape(state0)[-1]))
+    num_indices1 = int(np.log2(qml.math.shape(state1)[-1]))
```
Comment (timmysilv) [inline pennylane/math/quantum.py:843]: it seems you're just taking what already existed, but do we need the `log2` if we're just making sure they're equal?

**Review by timmysilv — APPROVED** [review]
🎉  also not sure about the name, but I'm fine with what's here

File: pennylane/math/quantum.py (hunk 1 above)
Comment (eddddddy) [reply pennylane/math/quantum.py:843]: Ah good catch

**Review by mudit2812 — APPROVED** [review]

**Comment by eddddddy** [comment]
@timmysilv I'm happy to consider alternative suggestions for the name `fidelity_statevector`, but @trbromley confirmed that it should be fine as is.

**Comment by eddddddy** [comment]
[sc-41139]

**Comment by timmysilv** [comment]
my comment was misleading now that I read it, I'm not picky on the name :shipit:

**Pull request merged**